Storage test infra refactor

jaschrep-msft opened the pull request:
Pulls some basics of client tests out of `XxxTestBase` hierarchy and into separate utility classes so they can be accessed outside of said hierarchy. Useful for writing unified tests across multiple Storage services without excessive code duplication.

Key Changes:
- `TenantConfigurationBuilder` houses code previously in `StorageTestBase` that grabs `TenantConfiguration`s out of a config file.
- `ClientBuilder` houses genericized code previously in `BlobTestBase`/`DataLakeTestBase`/etc. that creates `XxxServiceClient`s out of results from `TenantConfigurationBuilder`.
- Each service test suite has a `ClientBuilderExtensions` class that decorates the `ClientBuilder` with functionality that doesn't fit easy into generics.

Notes to reviewers:
- This is not intended to be a comprehensive change covering all facets of test infra. It is a starting point that will lead towards better maintenance and immediately open some doors for current testing needs.
- Currently, these separated classes are still dependent on some functionality provided by Azure Core's `RecordedTestBase`, and so we still have to pass in an instance of that. We need access to the `Recording` management and `InstrumentClient`. Not ideal but still a significant improvement.

jaschrep-msft committed:
Initial test resource refactor

jaschrep-msft committed:
converted blob/datalake/share to new test infra

jaschrep-msft committed:
ported test infra to queues

jaschrep-msft committed:
minor refactor

kasobol-msft commented on sdk/storage/Azure.Storage.Blobs/tests/BlobTestBase.cs:
```diff
@@ -36,35 +35,58 @@ namespace Azure.Storage.Test.Shared
+        /// <summary>
+        /// Source of clients.
+        /// </summary>
+        protected ClientBuilder<BlobServiceClient, BlobClientOptions> Clients { get; }
```
nit: By looking at name I'd guess that it's a `List<Client>`. I'd be verbose in property naming, with IDE support that shouldn't be a big deal.

kasobol-msft commented on sdk/storage/Azure.Storage.Common/tests/Shared/ClientBuilder.cs:
```diff
@@ -0,0 +1,178 @@
+            _getServiceClientOptions = getServiceClientOptions;
+        }
+
+        public TServiceClient GetServiceClient_SharedKey(TServiceClientOptions options = default)
+            => GetServiceClientFromSharedKeyConfig(Tenants.TestConfigDefault, options);
+
+        public TServiceClient GetServiceClient_SecondaryAccount_SharedKey()
+            => GetServiceClientFromSharedKeyConfig(Tenants.TestConfigSecondary);
+
+        public TServiceClient GetServiceClient_PreviewAccount_SharedKey()
+            => GetServiceClientFromSharedKeyConfig(Tenants.TestConfigPreviewBlob);
+
+        public TServiceClient GetServiceClient_PremiumBlobAccount_SharedKey()
+            => GetServiceClientFromSharedKeyConfig(Tenants.TestConfigPremiumBlob);
+        public TServiceClient GetServiceClient_OAuth() =>
+            GetServiceClientFromOauthConfig(Tenants.TestConfigOAuth);
+
+        public TServiceClient GetServiceClient_OAuthAccount_SharedKey() =>
+            GetServiceClientFromSharedKeyConfig(Tenants.TestConfigOAuth);
+
+        public TServiceClient GetServiceClient_ManagedDisk() =>
+            GetServiceClientFromSharedKeyConfig(Tenants.TestConfigManagedDisk);
+
+        public TServiceClient GetServiceClient_PremiumBlob() =>
+            GetServiceClientFromSharedKeyConfig(Tenants.TestConfigPremiumBlob);
+
+        public TServiceClient GetServiceClient_PremiumFile() =>
+            GetServiceClientFromSharedKeyConfig(Tenants.TestConfigPremiumFile);
+
+        public TServiceClient GetServiceClient_Hns() =>
+            GetServiceClientFromSharedKeyConfig(Tenants.TestConfigHierarchicalNamespace);
+
+        public TServiceClient GetServiceClient_SoftDelete() =>
+            GetServiceClientFromSharedKeyConfig(Tenants.TestConfigSoftDelete);
```
I think these should live in service specific packages. Why Queues would be coupled to HNS account for example?
Btw. instead of embedding all the tenants inside this class we could make `ClientBuilder(Tenant)` and then for example in the `BlobTestBase` have `ClientBuilder(Tenants.TestConfigDefault)`  ,  `ClientBuilder(Tenants.TestConfigSecondary)` and so on, multiple instances and each working with some tenant.

jaschrep-msft replied:
I've split these out into service-specific extensions.

I'm trying really hard not to scope creep this PR and so I'd rather not do the separate `ClientBuilder` per tenant just yet. I do think what we have now is still a noticeable improvement to our current infra.

kasobol-msft replied:
can you log work item for remaining part?

jaschrep-msft replied:
#24457

kasobol-msft commented on sdk/storage/Azure.Storage.Common/tests/Shared/ClientBuilder.cs:
```diff
@@ -0,0 +1,178 @@
+            GetServiceClientFromOauthConfig(Tenants.TestConfigOAuth);
+
+        public TServiceClient GetServiceClient_OAuthAccount_SharedKey() =>
+            GetServiceClientFromSharedKeyConfig(Tenants.TestConfigOAuth);
```
for the future perhaps. I think that separate "Oauth" tenant doesn't make sense, credentials is orthogonal concept to account feature set , i.e. I can connect with token credential to all of these account here (except files).
I.e. default tenant and oauth tenant can be the same thing.

jaschrep-msft replied:
Definitely a good idea but that is a huge scope creep for this PR.

kasobol-msft replied:
can you log work item for this?

jaschrep-msft replied:
#24458

kasobol-msft reviewed (commented):
very nice, this is going into right direction.

jaschrep-msft committed:
CI sanity check

jaschrep-msft committed:
Fixed default client options generation for tests

jaschrep-msft committed:
fix incorrect service uri bug

jaschrep-msft committed:
fixed default test tenant for datalake tests

jaschrep-msft committed:
Fixed cred supplier for datalake tests

jaschrep-msft committed:
split some functionality to specific services

ServiceCLient getters are now in package-specific extensions instead of
on the client builder directly.
Property rename.

kasobol-msft commented on sdk/storage/Azure.Storage.Common/tests/Shared/StorageTestBase.cs:
```diff
@@ -34,7 +34,7 @@ static StorageTestBase()
 
         public StorageTestBase(bool async, RecordedTestMode? mode = null)
-            : base(async, mode)
+            : base(async, RecordedTestMode.Playback)
```
ups.

kasobol-msft commented on sdk/storage/Azure.Storage.Blobs/tests/ClientBuilderExtensions.cs:
```diff
@@ -25,6 +25,33 @@ public static string GetNewBlockName(this BlobsClientBuilder clientBuilder)
+        public static BlobServiceClient GetServiceClient_ManagedDisk(this BlobsClientBuilder clientBuilder) =>
+            clientBuilder.GetServiceClientFromSharedKeyConfig(clientBuilder.Tenants.TestConfigManagedDisk);
+
+        public static BlobServiceClient GetServiceClient_Hns(this BlobsClientBuilder clientBuilder) =>
```
we don't need that here right?

jaschrep-msft replied:
Yes we do. There is one specific API we test that needs an HNS account.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs.Batch/tests/BlobBatchClientTests.cs:
```diff
@@ -16,6 +16,7 @@
 using Azure.Storage.Test.Shared;
 using Moq;
 using NUnit.Framework;
+using static Azure.Storage.Blobs.Tests.ClientBuilderExtensions;
```
Why is the import static?

jaschrep-msft replied:
This is just what visual studio decided to do when using alt+enter to import the missing stuff. Should we not be doing this?

seanmcc-msft replied:
If regular import works instead, we should use regular import for consistency.

jaschrep-msft committed:
cleanup

jaschrep-msft committed:
refactored disposing containers
